pubsub/pubsub_test.go

```diff
@@ -0,0 +1,397 @@
+						return
+					}
+					if res == nil {
+						continue
+					}
+					gotMessages[idx][res.ID] = res.Value
+					resp := fmt.Sprintf("result for: %v", res.ID)
+					if err := c.SetResult(ctx, res.ID, resp); err != nil {
+						t.Errorf("Error setting a result: %v", err)
+					}
+					wantResponses[idx] = append(wantResponses[idx], resp)
+				}
+			})
+	}
+	return gotMessages, wantResponses
+}
+
+func TestRedisClaimingOwnership(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	producer, consumers := newProducerConsumers(ctx, t)
+	producer.Start(ctx)
+	promises, err := produceMessages(ctx, producer)
+	if err != nil {
+		t.Fatalf("Error producing messages: %v", err)
+	}
+
+	// Consumer messages in every third consumer but don't ack them to check
+	// that other consumers will claim ownership on those messages.
+	for i := 0; i < len(consumers); i += 3 {
+		i := i
+		if _, err := consumers[i].Consume(ctx); err != nil {
+			t.Errorf("Error consuming message: %v", err)
+		}
+		consumers[i].StopAndWait()
+	}
+
+	gotMessages, wantResponses := consume(ctx, t, consumers, 3)
+	gotResponses, err := awaitResponses(ctx, promises)
+	if err != nil {
+		t.Fatalf("Error awaiting responses: %v", err)
+	}
+	for _, c := range consumers {
+		c.StopWaiter.StopAndWait()
+	}
+	got, err := mergeValues(gotMessages)
+	if err != nil {
+		t.Fatalf("mergeMaps() unexpected error: %v", err)
+	}
+	want := wantMessages(messagesCount)
+	if diff := cmp.Diff(want, got); diff != "" {
+		t.Errorf("Unexpected diff (-want +got):\n%s\n", diff)
+	}
+	wantResp := flatten(wantResponses)
+	sort.Strings(gotResponses)
+	if diff := cmp.Diff(wantResp, gotResponses); diff != "" {
+		t.Errorf("Unexpected diff in responses:\n%s\n", diff)
+	}
+	if cnt := producer.promisesLen(); cnt != 0 {
+		t.Errorf("Producer still has %d unfullfilled promises", cnt)
+	}
+}
+
+func TestRedisClaimingOwnershipReproduceDisabled(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	producer, consumers := newProducerConsumers(ctx, t, &disableReproduce{})
+	producer.Start(ctx)
+	promises, err := produceMessages(ctx, producer)
+	if err != nil {
+		t.Fatalf("Error producing messages: %v", err)
+	}
+
+	// Consumer messages in every third consumer but don't ack them to check
+	// that other consumers will claim ownership on those messages.
+	for i := 0; i < len(consumers); i += 3 {
+		i := i
+		if _, err := consumers[i].Consume(ctx); err != nil {
+			t.Errorf("Error consuming message: %v", err)
+		}
+		consumers[i].StopAndWait()
+	}
+
+	gotMessages, _ := consume(ctx, t, consumers, 3)
+	gotResponses, err := awaitResponses(ctx, promises)
+	if err == nil {
+		t.Fatalf("All promises were fullfilled with reproduce disabled and some consumers killed")
+	}
+	for _, c := range consumers {
+		c.StopWaiter.StopAndWait()
+	}
+	got, err := mergeValues(gotMessages)
+	if err != nil {
+		t.Fatalf("mergeMaps() unexpected error: %v", err)
+	}
+	wantMsgCnt := messagesCount - (consumersCount / 3) - (consumersCount % 3)
```

Comment:
I don't think that's correct? e.g. for consumersCount = 2 you want 1 and not 2 messages.
I think it should be: messagesCount - ((messagesCount + 2) /3)